lib/ecto/changeset/relation.ex

```diff
@@ -83,7 +83,8 @@ defmodule Ecto.Changeset.Relation do
   def cast(%{cardinality: :many} = relation, params, current, on_cast) when is_map(params) do
     params =
       params
-      |> Enum.sort_by(&elem(&1, 0))
+      |> Enum.map(&key_as_int/1)
+      |> Enum.sort
```

Comment:
I think it should be `Enum.sort_by(&elem(&1, 0))` istead of `Enum.sort`.
otherwise it will sort by values as well which is unnecessary (and slower), e.g.:

```elixir
iex> Enum.sort([%{"1" => "b"}, %{"1" => "a"}])
[%{"1" => "a"}, %{"1" => "b"}]
```